Group bar charts by project

XD-182 opened the pull request:
Cleaned the Files up by removing whitespace differences. New Pull Request because my old Mac won't cooperate, apologies.

XD-182 committed:
Cleaned Diff version of AmountsChart and ReportView with corrected file structure

jp-hoehmann commented on Chronos/Views/AmountsChart.swift:
```diff
@@ -11,6 +11,14 @@ import SwiftUI
 import ObservedOptionalObject
 
+/// Helper structure for sorting data
+
```
Remove extra blank line

jp-hoehmann replied:
Remove extra blank line

jp-hoehmann commented on Chronos/Views/AmountsChart.swift:
```diff
@@ -11,6 +11,14 @@ import SwiftUI
 
+/// Helper structure for sorting data
+
+struct AggregatedEntry: Identifiable {
```
If this is a helper struct only needed in AmountsChart, it should be part of amounts chart and marked private (in this project, it should be at the top of the struct that its in, above the life cycle methods).

jp-hoehmann commented on Chronos/Views/AmountsChart.swift:
```diff
@@ -125,23 +133,54 @@ struct AmountsChart: View {
+    private var groupedAggregatedEntries: [Date: [AggregatedEntry]] {
+        Dictionary(grouping: aggregatedEntries, by: { $0.day })
+    }
+    /// Modified chart to use aggregated data
```
This should have been a commit message.

jp-hoehmann commented on Chronos/Views/AmountsChart.swift:
```diff
@@ -203,3 +242,4 @@ class AmountsChart_Previews: PreviewProvider {
     }
 
 }
+
```
Remove extra blank line.

jp-hoehmann commented on Chronos/Views/AmountsChart.swift:
```diff
@@ -11,6 +11,14 @@ import SwiftUI
+    let id = UUID()
+    let day: Date
+    let project: String
+    let totalHours: Double
```
Prefer type-safe code. This should be a `TimeInterval` (and should thus be renamed to something like `totalTime`).

jp-hoehmann commented on Chronos/Views/AmountsChart.swift:
```diff
@@ -11,6 +11,14 @@ import SwiftUI
+struct AggregatedEntry: Identifiable {
+    let id = UUID()
+    let day: Date
+    let project: String
```
Since this isn't actually a `Project` and in fact might not even be the name of a project (since it can be `"Other"`), it should probably be renamed to something like `segmentName`.

jp-hoehmann commented on Chronos/Views/AmountsChart.swift:
```diff
@@ -125,23 +133,54 @@ struct AmountsChart: View {
-                                                .map(String.init)
-                                                ?? (project != nil ? project!.name : "Other"))
+        let sortedDays = groupedAggregatedEntries.keys.sorted()
+        let calendar = Calendar.current
```
I don't think the `Calendar` here is used for anything. Please remove it.

jp-hoehmann commented on Chronos/Views/AmountsChart.swift:
```diff
@@ -125,23 +133,54 @@ struct AmountsChart: View {
+                .split(separator: Project.pathSeparator)
+                .first
+                .map(String.init)
+                ?? (project != nil ? project!.name : "Other")
```
This can be written as `?? project?.name ?? "Other"`, which is more readable.

jp-hoehmann commented on Chronos/Views/AmountsChart.swift:
```diff
@@ -125,23 +133,54 @@ struct AmountsChart: View {
+    private var aggregatedEntries: [AggregatedEntry] {
+        let calendar = Calendar.current
+        let groups = Dictionary(grouping: entries) { entry -> DayProjectKey in
+            let day = calendar.startOfDay(for: entry.start)
```
This can be written as `let day = entry.start.day`, which is more readable.

jp-hoehmann reviewed (changes requested):
I like your solution, you've already found most of the edge-cases and handled them correctly. I also like your use of functional style, which keeps the parts of your code that do the heavy lifting concise and readable for the most part.

Your solution currently falls apart when a bar represents more than a day. Please make sure that the chart aggregates the entries correctly when each bar represents a week, a month, or a quarter.

Right now, your solution is also unnecessarily complicated. Do you think you can simplify the code a little?

XD-182 commented:
Hi, I have been pretty busy with school as it’s the week of the year where I have my most important exams. Hence I’ll probably only get to take a closer look at the requested changes on Thursday, as I’ll have some more free time moving forward from then. I apologize for the delay!

jp-hoehmann commented:
No worries, we're not in a hurry. Good luck with your exams! :-)

XD-182 committed:
Removed Extra blank lines, Renamed project to segmentName for readability.

XD-182 committed:
Moved AmountsChart to a better location and made it private

XD-182 committed:
Revamped totalHours to use TimeInterval and renamed it to totalTime. Removed line of code that was never used.

XD-182 committed:
Changed the aggregation to use a helper func "truncatedDate" which facilitates proper sorting when a bar represents different time intervals (month).

XD-182 committed:
Did some simple renaming to reflect the new working mechanisms of the code and thus improve readability.

XD-182 commented:
I made some small simplifications as requested, implemented the changes you suggested and found a way to solve the issue with the bar not having been ordered correctly when it represented more than a day. Additionally I clarified the code by renaming some elements, as the working mechanisms have changed a bit. Let me know if my commit frequency is now up to your standards.

jp-hoehmann commented:
Your commits look good now. You missed a requested change, though. Also maybe read over your code again before re-requesting review, just to make there isn't any obvious formatting or style issues.

XD-182 committed:
Fixed some formatting and renamed day to unitDate for readability.

XD-182 commented:
Apologies for the oversight, I fixed some formatting issues I missed when proofreading by removing unnecessary blank lines. However I'm not certain what you're referring to when you pointed out that I missed a requested change. The only thing I didn't implement was 
> This can be written as `let day = entry.start.day`, which is more readable. 

This however doesn't apply anymore because my current solution doesn't use the built‑in API—Calendar’s startOfDay(for:) anymore so I'm not sure it would work. Let me know if I missed something!

jp-hoehmann commented:
I meant this one:

https://github.com/valomedia/Pandatrack/pull/5/files#r1991155177

XD-182 commented:
I actually registered that particular comment of yours as advice regarding my message habits rather than you asking me to act and add a commit message retrospectively, sorry I got that wrong. Would you want me to do something retrospectively regarding this?

jp-hoehmann commented:
Ah, yes, sorry. Please remove the comment.

XD-182 committed:
Removed unnecessary comment, minor form correction

jp-hoehmann reviewed (changes requested):
Looks good. Do you think you can simplify it a little more? I don't mean from a code-style perspective – you have mentioned that you haven't used Swift before, so where there was a nicer way to write things, I've just pointed it out directly – but rather making what the code does simpler.

Right now, your code contains a bunch of needless complexity where things can be either dropped because they don't really accomplish anything, or where you do things that you end up undoing later, just to do them over again. If you don't see it, I can give you more hints, but maybe you can find some opportunities for improvement yourself, if you mentally go over the steps that are actually needed and the steps your code takes?

XD-182 committed:
Refactored aggregation and chart logic by eliminating one unnecessary mapping step and consolidating all grouping logic into groupedAggregatedEntries, reducing conceptual complexity

XD-182 commented:
I noticed a bit of redundancy within the aggregation logic. Basically I cut one mapping step from the process. this seems to not have reduced the line count but it does reduce conceptual complexity which I suspect you're after? I suspect there are some opportunities for cutting a few lines left here (maybe combining Grouping Steps or replacing Dictionary(grouping:by:) with reduce(into:)), but I think the would reduce readability a lot in exchange for minimal reduction and in any case I'm uncertain if these are viable implementations. Anyways this is my attempt at simplification, if this isn't what you were referring to I might need a hint.

jp-hoehmann commented on Chronos/Views/AmountsChart.swift:
```diff
@@ -125,40 +141,68 @@ struct AmountsChart: View {
+    struct UnitSegmentKey: Hashable {
+        let unitDate: Date
+        let segmentName: String
+    }
```
Your current solution doesn't need this anymore, please remove.

jp-hoehmann commented on Chronos/Views/AmountsChart.swift:
```diff
@@ -125,40 +141,68 @@ struct AmountsChart: View {
-                                                .first
-                                                .map(String.init)
-                                                ?? (project != nil ? project!.name : "Other"))
+        let sortedDates = groupedAggregatedEntries.keys.sorted()
```
I don't think this sorting step does anything. Please remove it.

jp-hoehmann commented on Chronos/Views/AmountsChart.swift:
```diff
@@ -125,40 +141,68 @@ struct AmountsChart: View {
+            Chart {
+                ForEach(sortedDates, id: \.self) { date in
+                    // For each unit date, sort the segments by total time
+                    let segments = groupedAggregatedEntries[date]!.sorted { $0.totalTime > $1.totalTime }
```
With your current solution, this can just go into the logic for groupedAggregatedEntries, making this bit a little cleaner.

jp-hoehmann reviewed (changes requested):
You are absolutely right, I don't care about line count, as long as it's readable :-)

Your latest commit again changes a bunch of stuff in a single commit, most of the changes are related to renaming things to reflect that we have variable segments now, instead of days. Please make sure to make two commits in the future.

With your latest simplification, you have reverted from the nice, functional style this solution had to writing imperative code. Can you rewrite the logic for generating the `groupedAggregatedEntries` in functional style?

Also I should maybe say, that we're just going back and forth like this because this is an exercise and I wanna figure out what the limits of your capabilities are. So don't worry that I'm asking you to change it around so much. Doesn't mean there's something wrong with your code :-)

XD-182 committed:
Removed Unnecessary code

XD-182 committed:
Moved sorting for segments into GroupedAggregatedEntries

XD-182 committed:
Minor change to groupedAggregatedEntries which makes the style more functional thus improving readability.

XD-182 commented:
It seems my biggest difficulty is judging when to commit, I suspect I'll get a feeling for that over time. That aside, thank you for clarifying the reason behind the back and forth, I was beginning to think my coding skills were entirely inadequate :-)

I implemented the simplifications you mentioned and also did some restyling of groupedAggregatedEntries to make it more functional. I believe it does fit the functional style of the remaining code now that I altered it a bit, but let me know what you think. It definitely improves on readability though even though it's only a minor-ish change.

jp-hoehmann commented:
Yeah, don't worry about that. That we've been going for this long actually means I think you're quite good, so I gotta keep pressing to see what you can and can't do :-)

Your solution is more functional now, but still a bit messy, with the creation of a Dictionary, which is then turned into an Array of Tuples, just to be turned back into a Dictionary in the end. There is a function called `Dictionary.mapValues()` that can probably make this more readable. Can you figure out how?

XD-182 committed:
Added custom mapValuesWithKey extension to Dictionary, allowing transformation of each key-value pair with a closure that receives both the key and the value. This extension is needed because the standard mapValues only provides the value, which limits transformations that require the key.

XD-182 committed:
Adjusted groupedAggregatedEntries in regards to the new extension.

XD-182 commented:
Thank you, I'm glad to hear I'm doing good so far even though I'm kinda new to Swift.

I did figure out a way to do what you asked but I ended up having to create a new extension as you can see in my recent commits. This makes the code longer but I think it improves conceptual simplicity and readability nonetheless. If my solution is not what you were thinking of, I think there's another way to do something similar by using reduce(into:) (could be shorter, probably less functional though) but that's just an idea.

jp-hoehmann commented on Chronos/Views/AmountsChart.swift:
```diff
@@ -11,10 +11,26 @@ import SwiftUI
 
 struct AmountsChart: View {
+    
+    // MARK: - Private Helper Structures
```
Please remove this mark.

jp-hoehmann commented on Chronos/Views/AmountsChart.swift:
```diff
@@ -11,10 +11,26 @@ import SwiftUI
+            let unitDate: Date
+            let segmentName: String
+            let totalTime: TimeInterval
+        }
```
Fix indentation

jp-hoehmann commented on Chronos/Views/AmountsChart.swift:
```diff
@@ -11,10 +11,26 @@ import SwiftUI
+            return Calendar.current.startOfDay(for: date)
+        }
+        return interval.start
+    }
```
This seems like it could be useful in other parts of the project, can you turn it into a public extension for Date?

jp-hoehmann reviewed (changes requested):
Yeah, I think the readability improvement from that last change isn't worth having to extend Dictionary. Can you revert the last commit?

Other than that, I've added a few more comments about little changes to make before we can merge this.

XD-182 committed:
Undid the previous two commits.

XD-182 committed:
Made truncatedDate a public extension. Adjusted the aggregation loic accordingly.

XD-182 commented:
Implemented everything you suggested. Let me know if it's ready or if more changes are needed.

jp-hoehmann commented on Chronos/Views/AmountsChart.swift:
```diff
@@ -11,10 +11,31 @@ import SwiftUI
 
 struct AmountsChart: View {
+    
+    ///Helper struct for sorting data
```
Please format this the same way the other doc-comments are formatted.

jp-hoehmann commented on Chronos/Views/AmountsChart.swift:
```diff
@@ -11,10 +11,31 @@ import SwiftUI
 import Charts
 import ObservedOptionalObject
 
+public extension Date {
+    /// Returns the start of the interval for the specified calendar component.
+    /// - Parameters:
+    ///   - component: The calendar component (e.g., .day, .weekOfYear, .month, etc.) to truncate to.
+    ///   - calendar: The calendar to use; defaults to Calendar.current.
+    /// - Returns: The start of the interval if available, otherwise the start of the day.
+    func truncated(to component: Calendar.Component, calendar: Calendar = .current) -> Date {
+        guard let interval = calendar.dateInterval(of: component, for: self) else {
+            return calendar.startOfDay(for: self)
+        }
+        return interval.start
+    }
+}
```
This should have its own file under `Extenisons`, similar to all the other extensions.

jp-hoehmann reviewed (changes requested):
I've made two more comments. Also there is one request from previous rounds that's still open. Is there a particular reason for that, or have you just missed it?

XD-182 committed:
Put Date Truncation in its own file and removed unnessecary blank line

XD-182 commented:
I implemented the changes you've requested, sorry for the wait. As for the Change I didn't implement, I assume you are referring to the removal of

>        let sortedDates = groupedAggregatedEntries.keys.sorted()

I didn't remove this on purpose, as without, my compiler is unable to type-check in reasonable time. Might just be my old hardware if it worked for you or something else, I didn't look into this much further.

jp-hoehmann commented on Chronos/Extensions/Date+Truncation.swift:
This file should be named the same as the extension it provides, so `Date+truncated`.

jp-hoehmann reviewed (changes requested):
I've made another comment. Also please make sure that you address all the ones that are already there (and when you do, maybe check if the same change is needed elsewhere), I know that these little code-style changes are annoying, but on real projects the customers often have their own people that look at the code, and if the formatting is messy, that just looks unprofessional. Also, can you please remove that extra sorting step? It makes that whole bit a bunch more messy than it needs to be.

The Swift compiler (like basically all compilers) is deterministic. If it doesn't compile for you, it doesn't compile for me. If you have a slower machine, that just means the “reasonable time” becomes longer. If your expression fails to type check, that almost always means you screwed up somewhere and the compiler failed to find the correct types because no correct types exist. If you ever do run into a case where it's really just that complex, the solution is to explicitly specify the types, so the compiler doesn't have to figure them out.

XD-182 committed:
Renaming Date+truncated to better reflect extension function

XD-182 committed:
Reformatted comment to look like other doc-comments

XD-182 committed:
Removed extra sorting step from Chart, now all the sorting is being handled by groupedAggregatedEntries making the code cleaner

XD-182 commented:
I did the reworks you requested. I found a way to remove the sorting step from chart, but I ended up having to change more logic than expected. Nevertheless I think the way it is now makes it more readable by a lot (Separation of Concerns, Improved Maintainability, add to that Compiler Friendliness). Let me know what you think.

jp-hoehmann commented on Chronos/Extensions/Date+truncated.swift:
```diff
@@ -0,0 +1,23 @@
+//
+//  Date+Truncation.swift
```
This line is wrong.

jp-hoehmann commented on Chronos/Extensions/Date+truncated.swift:
```diff
@@ -0,0 +1,23 @@
+import Foundation
+
+public extension Date {
+    /// Returns the start of the interval for the specified calendar component.
+    /// - Parameters:
+    ///   - component: The calendar component (e.g., .day, .weekOfYear, .month, etc.) to truncate to.
+    ///   - calendar: The calendar to use; defaults to Calendar.current.
+    /// - Returns: The start of the interval if available, otherwise the start of the day.
```
Make the formatting of the doc-comments match the rest of the project

jp-hoehmann commented on Chronos/Views/AmountsChart.swift:
```diff
@@ -11,10 +11,16 @@ import SwiftUI
 import Charts
 import ObservedOptionalObject
 
-
 // MARK: AmountsChart
 
+/// A helper struct for sorting data.
```
Make the formatting of the doc-comments match the rest of the project.

jp-hoehmann reviewed (changes requested):
The extra sorting step is still there, now you've just moved it elsewhere. Is there any reason you need to sort these? If not, please just remove the extra sorting step.

XD-182 committed:
Removed unnecessary outer sort from aggregated entries and reformatted comment.

jp-hoehmann reviewed (changes requested):
Most of the mess you've added trying to move the pointless sorting step is still there, please clean it up. Also make sure to actually address all the changes requested, before re-requesting review. This back-and-forth becomes way to time-consuming otherwise.

XD-182 commented:
I did some reworking of the Aggregation logic to make it simpler. this means creating some helper functions which I implemented below chart. I also implemented the changes you asked for Date+truncated, I believe every change request is now addressed. Let me know if you would like me to revert the changes I made to amounts chart as it's quite a substantial change, at least layout wise.

jp-hoehmann reviewed (changes requested):
Sorry, but no. This just isn't helpful whatsoever. All I was asking was for you to remove some pointless code that did nothing. Instead, you've for some reason introduced a helper struct, that as far as I can tell serves no purpose whatsoever, renamed `groupedAggregatedEntries` to `dailyAggregatedData`, which is just plain wrong, as the aggregation isn't daily, moved a bunch of stuff around, making this a pain to review, and above all, you *still* haven't removed that extra sorting step. Why?

Also it doesn't compile right now, so there's that. Please also fix that, so I can actually try it, before we go to merge.

XD-182 committed:
Cleaned up chart and sortedAggregatedEntries.
sortedAggregatedEntries:
• Removed an unnecessary variable by directly returning the mapped dictionary.
• Used cleaner inline grouping and mapping with a guard-like pattern in the closure.
chart:
• Extracted the day count into a variable for improved clarity.
• Adjusted indentation and grouping to make the view builder structure easier to follow without modifying its behavior.

XD-182 committed:
Somehow my Extension got lost so I recreated it???

XD-182 committed:
Made it use my new (old) extension

XD-182 commented:
I reverted, cleaned and force pushed the code. Somehow during that process I lost my extension so I recreated it. Now it's working though. I'll commit my project file in case it solves the compiling problem as well.

XD-182 committed:
project file in case its needed

XD-182 commented:
I apologize for any oversights that might have happened, to be honest it's been a bit of a rough time trying to revert all this, as it is my first time doing something like this.

jp-hoehmann reviewed (changes requested):
Ok, now we have two extensions on Date that seem to do exactly the same thing, the extra sorting step is still there and you've introduced new changes that weren't in any of your previous commits as far as I can tell.

It is really important that you read what you're asking me to review, before you tell me to review it. Otherwise this is just a waste of both of our time.
